dig/dig.go

```diff
@@ -0,0 +1,292 @@
+)
+
+var (
+	errParamType   = errors.New("graph injection must be done through a pointer or a function")
+	errReturnCount = errors.New("constructor function must return exactly one value")
+	errReturnKind  = errors.New("constructor return type must be a pointer")
+	errArgKind     = errors.New("constructor arguments must be pointers")
+)
+
+// New returns a new Dependency Injection Graph
+func New() Graph {
+	return newGraph()
+}
+
+// Graph facilitates basic dependency injection
+//
+// TODO: Register functions should take options to control resolve behaviour
+// for example, at the moment everything is a singleton and we need an option
+// to initialize a new instance on every resolve
+type Graph interface {
+	// Register into the dependency graph
+	// Parameter must be a pointer or a constructor function returning exactly one pointer
+	Register(interface{}) error
+
+	// Same as calling .Register for each provided vararg
+	// Returns the first error encountered
+	RegisterAll(...interface{}) error
+
+	// Resolve the dependencies of the object and populate the pointer value
+	Resolve(interface{}) error
+
+	// Reset the graph by removing all the registered nodes
+	Reset()
+}
+
+// Register an object in the dependency graph
+//
+// Provided argument must be a function that returns exactly one pointer argument
+// All arguments to the function must be pointers
+func (g *graph) Register(c interface{}) error {
+	ctype := reflect.TypeOf(c)
+
+	switch ctype.Kind() {
+	case reflect.Func:
+		if ctype.NumOut() != 1 {
+			return errReturnCount
+		}
+
+		objType := ctype.Out(0)
+		if objType.Kind() != reflect.Ptr && objType.Kind() != reflect.Interface {
+			return errReturnKind
+		}
+
+		return g.registerConstructor(c)
+	case reflect.Ptr:
+		return g.registerObject(c)
+	default:
+		return errParamType
+	}
+}
+
+// Resolve all of the dependencies of the provided class
+//
+// Provided object must be a pointer
+// Any dependencies of the object will receive constructor calls, or be initialized (once)
+// Constructor with return value *object will be called
+//
+// TODO(glib): catch any and all panics from this method, as there is a lot of reflect going on
+func (g *graph) Resolve(obj interface{}) error {
+	objType := reflect.TypeOf(obj)
+	if objType.Kind() != reflect.Ptr {
+		return fmt.Errorf("can not resolve non-pointer object of type %v", objType)
+	}
+
+	objElemType := reflect.TypeOf(obj).Elem()
+	objVal := reflect.ValueOf(obj)
+
+	// check if the type is a registered node
+	n, ok := g.nodes[objElemType]
+	if !ok {
+		return fmt.Errorf("type %v is not registered", objType)
+	}
+
+	// check that all the dependencies have nodes present in the graph
+	// doesn't mean everything will go smoothly during resolve, but it
+	// drastically increases the chances that we're not missing something
+	for _, node := range g.nodes {
+		for _, dep := range node.dependencies() {
+			// check that the dependency is a registered node
+			if _, ok := g.nodes[dep]; !ok {
```

Comment:
Can the dependency register checks go inside the n.value function itself? Makes sense to check this every time value is called. Along with that, you can remove the dependencies function from the object interface since value function is the only one that needs access to it.